File: src/lex/lex.zig
pub const Lexer = struct {
    src: []const u8,
    cursor: Index,

    // todo
    err: void,

    const Self = @This();

    pub fn next(self: *Self) Token {
        const b = self.src;
        var old_cursor = self.cursor;

        while (true) : (self.cursor += 1) {
            if (self.cursor >= self.src.len) return token(.eof, 0, 0);
            const c = b[self.cursor];

            switch (c) {
                '^' => {
                    self.cursor += 1;
                    return token(.@"^", old_cursor, self.cursor);
                },
                '@' => {
                    self.cursor += 1;
                    return token(.@"@", old_cursor, self.cursor);
                },
                '#' => {
                    self.cursor += 1;
                    return token(.@"#", old_cursor, self.cursor);
                },
                '|' => {
                    self.cursor += 1;
                    return token(.@"|", old_cursor, self.cursor);
                },
                '\\' => {
                    self.cursor += 1;
                    return token(.@"\\", old_cursor, self.cursor);
                },
                '&' => {
                    self.cursor += 1;
                    return token(.@"&", old_cursor, self.cursor);
                },
                '(' => {
                    self.cursor += 1;
                    return token(.@"(", old_cursor, self.cursor);
                },
                ')' => {
                    self.cursor += 1;
                    return token(.@")", old_cursor, self.cursor);
                },
                '[' => {
                    self.cursor += 1;
                    return token(.@"[", old_cursor, self.cursor);
                },
                ']' => {
                    self.cursor += 1;
                    return token(.@"]", old_cursor, self.cursor);
                },
                ';' => {
                    self.cursor += 1;
                    return token(.@";", old_cursor, self.cursor);
                },
                ',' => {
                    self.cursor += 1;
                    return token(.@",", old_cursor, self.cursor);
                },
                '?' => {
                    self.cursor += 1;
                    return token(.@"?", old_cursor, self.cursor);
                },
                '~' => {
                    if (self.eatChar('>')) {
                        self.cursor += 1;
                        return token(.@"~>", old_cursor, self.cursor);
                    } else {
                        self.cursor += 1;
                        return token(.@"~", old_cursor, self.cursor);
                    }
                },
                '>' => {
                    if (self.seperated(self.cursor)) {
                        self.cursor += 2;
                        return token(.@" > ", old_cursor, self.cursor);
                    } else {
                        if (self.eatChar('=')) {
                            self.cursor += 1;
                            return token(.@">=", old_cursor, self.cursor);
                        } else {
                            self.cursor += 1;
                            return token(.@">", old_cursor, self.cursor);
                        }
                    }
                },

                '<' => {
                    if (self.seperated(self.cursor)) {
                        self.cursor += 2;
                        return token(.@" < ", old_cursor, self.cursor);
                    } else {
                        if (self.eatChar('=')) {
                            self.cursor += 1;
                            return token(.@"<=", old_cursor, self.cursor);
                        } else {
                            self.cursor += 1;
                            return token(.@"<", old_cursor, self.cursor);
                        }
                    }
                },

                '+' => {
                    if (self.seperated(self.cursor)) {
                        self.cursor += 2;
                        return token(.@" + ", old_cursor, self.cursor);
                    } else {
                        if (self.eatChar('=')) {
                            self.cursor += 1;
                            return token(.@"+=", old_cursor, self.cursor);
                        } else if (self.eatChar('+')) {
                            self.cursor += 1;
                            return token(.@"++", old_cursor, self.cursor);
                        } else {
                            self.cursor += 1;
                            return token(.@"+", old_cursor, self.cursor);
                        }
                    }
                },
                '-' => {
                    if (self.seperated(self.cursor)) {
                        self.cursor += 2;
                        return token(.@" - ", old_cursor, self.cursor);
                    } else {
                        if (self.eatChar('=')) {
                            self.cursor += 1;
                            return token(.@"-=", old_cursor, self.cursor);
                        } else if (self.eatChar('>')) {
                            self.cursor += 1;
                            return token(.@"->", old_cursor, self.cursor);
                        } else if (self.eatChar('-')) {
                            // comment
                            while (true) : (self.cursor += 1) {
                                if (self.cursor >= self.src.len) return token(.comment, old_cursor, self.cursor - 1);
                                const c_ = b[self.cursor];
                                if (c_ == '\n') {
                                    self.cursor += 1;
                                    return token(.comment, old_cursor, self.cursor);
                                }
                            }
                        } else {
                            self.cursor += 1;
                            return token(.@"-", old_cursor, self.cursor);
                        }
                    }
                },

                '!' => {
                    if (self.eatChar('=')) {
                        self.cursor += 1;
                        return token(.@"!=", old_cursor, self.cursor);
                    } else {
                        self.cursor += 1;
                        return token(.@"!", old_cursor, self.cursor);
                    }
                },

                '=' => {
                    if (self.eatChar('=')) {
                        self.cursor += 1;
                        return token(.@"==", old_cursor, self.cursor);
                    } else if (self.eatChar('>')) {
                        self.cursor += 1;
                        return token(.@"=>", old_cursor, self.cursor);
                    } else {
                        self.cursor += 1;
                        return token(.@"=", old_cursor, self.cursor);
                    }
                },
                '.' => {
                    if (self.eatChar('.')) {
                        if (self.eatChar('.')) {
                            self.cursor += 1;
                            return token(.@"...", old_cursor, self.cursor);
                        } else if (self.eatChar('=')) {
                            self.cursor += 1;
                            return token(.@"..=", old_cursor, self.cursor);
                        } else {
                            self.cursor += 1;
                            return token(.@"..", old_cursor, self.cursor);
                        }
                    } else {
                        self.cursor += 1;
                        return token(.@".", old_cursor, self.cursor);
                    }
                },
                '%' => {
                    if (self.seperated(self.cursor)) {
                        self.cursor += 2;
                        return token(.@" % ", old_cursor, self.cursor);
                    } else {
                        if (self.eatChar('=')) {
                            self.cursor += 1;
                            return token(.@"%=", old_cursor, self.cursor);
                        } else {
                            self.cursor += 1;
                            return token(.@"%", old_cursor, self.cursor);
                        }
                    }
                },
                '{' => {
                    if (self.eatChar('-')) {
                        // comment
                        while (true) : (self.cursor += 1) {
                            if (self.cursor >= self.src.len) return token(.comment, old_cursor, self.cursor - 1);
                            const c_ = b[self.cursor];
                            if (c_ == '-') {
                                if (self.eatChar('}')) {
                                    self.cursor += 1;
                                    return token(.comment, old_cursor, self.cursor);
                                }
                            }
                        }
                    } else {
                        self.cursor += 1;
                        return token(.@"{", old_cursor, self.cursor);
                    }
                },
                '}' => {
                    self.cursor += 1;
                    return token(.@"}", old_cursor, self.cursor);
                },
                ':' => {
                    if (self.eatChar('-')) {
                        self.cursor += 1;
                        return token(.@":-", old_cursor, self.cursor);
                    } else if (self.eatChar(':')) {
                        self.cursor += 1;
                        return token(.@"::", old_cursor, self.cursor);
                    } else {
                        self.cursor += 1;
                        return token(.@":", old_cursor, self.cursor);
                    }
                },
                '*' => {
                    if (self.seperated(self.cursor)) {
                        self.cursor += 2;
                        return token(.@" * ", old_cursor, self.cursor);
                    } else {
                        self.cursor += 1;
                        return token(.@"*", old_cursor, self.cursor);
                    }
                },
                '/' => {
                    if (self.seperated(self.cursor)) {
                        self.cursor += 2;
                        return token(.@" / ", old_cursor, self.cursor);
                    } else {
                        self.cursor += 1;
                        return token(.@"/", old_cursor, self.cursor);
                    }
                },
                'a'...'z', 'A'...'Z' => return self.recognizeId(),

                '`' => return self.recognizeArbitaryId(),

                '_' => unreachable,

                '0'...'9' => return self.recognizeDigit(),

                ' ', '\t', '\n' => {
                    old_cursor += 1;
                },

                '"' => return self.recognizeStr(),

                else => unreachable,
            }
        }

        return token(.eof, 0, 0);
    }
    pub fn init(src: []const u8) Self {
        return .{
            .src = src,
            .cursor = 0,
            .err = undefined,
        };
    }
    pub fn deinit(_: *Self) void {}

    fn recognizeStr(self: *Self) Token {
        const b = self.src;
        const old_cursor = self.cursor;
        var started: bool = false;

        while (true) : (self.cursor += 1) {
            if (self.cursor >= self.src.len) return token(.eof, 0, 0);
            const c = b[self.cursor];

            switch (c) {
                '"' => {
                    if (!started)
                        started = true
                    else if (self.cursor - 1 >= 0 and (b[self.cursor - 1] != '\\')) {
                        self.cursor += 1;
                        return token(.str, old_cursor, self.cursor);
                    }
                },
                else => {},
            }
        }
        return token(.invalid, old_cursor, self.cursor);
    }
    fn recognizeMultiLineStr() void {}
    fn recognizeComment() void {}

    // BUG: `some`，当`为最后一个字符时，len不包括其在内
    fn recognizeId(self: *Self) Token {
        const b = self.src;
        const old_cursor = self.cursor;
        while (true) : (self.cursor += 1) {
            if (self.cursor >= self.src.len)
                if (Token.keywords.get(b[old_cursor..self.cursor])) |kw|
                    return token(kw, old_cursor, self.cursor)
                else {
                    return token(.id, old_cursor, self.cursor);
                };
            const c = b[self.cursor];

            switch (c) {
                'a'...'z', 'A'...'Z', '0'...'9', '_' => {},
                else => {
                    if (Token.keywords.get(b[old_cursor..self.cursor])) |kw|
                        return token(kw, old_cursor, self.cursor)
                    else {
                        return token(.id, old_cursor, self.cursor);
                    }
                },
            }
        }
    }
    fn recognizeDigit(self: *Self) Token {
        const b = self.src;
        const old_cursor = self.cursor;
        var met_dot: bool = false;

        while (true) : (self.cursor += 1) {
            if (self.cursor >= self.src.len) return token(if (met_dot) .real else .int, old_cursor, self.cursor);
            const c = b[self.cursor];

            switch (c) {
                '0'...'9' => {},
                '.' => {
                    if (!met_dot) {
                        met_dot = true;
                    } else {
                        // ERR, todo
                    }
                },
                else => {
                    return token(if (met_dot) .real else .int, old_cursor, self.cursor);
                },
            }
        }
    }
    fn recognizeArbitaryId(self: *Self) Token {
        const b = self.src;
        const old_cursor = self.cursor;
        var started: bool = false;

        while (true) : (self.cursor += 1) {
            if (self.cursor >= self.src.len) return token(.eof, 0, 0);
            const c = b[self.cursor];

            switch (c) {
                '`' => {
                    if (!started)
                        started = true
                    else {
                        self.cursor += 1;
                        return token(.id, old_cursor, self.cursor);
                    }
                },
                else => {},
            }
        }
        return token(.invalid, old_cursor, self.cursor);
    }

    fn peak(self: Self, char: u8) bool {
        const at = self.cursor;
        return if (self.src.len <= at + 1) false else if (self.src[at + 1] == char) true else false;
    }

    fn seperated(self: Self, at: Index) bool {
        return if (self.src.len <= at + 1 or at < 1) false else if (self.src[at + 1] == ' ' and self.src[at - 1] == ' ') true else false;
    }

    fn eatChar(self: *Self, char: u8) bool {
        var result: bool = false;

        if (self.src.len <= self.cursor + 1) return false;

        if (self.src[self.cursor + 1] == char) {
            self.cursor += 1;
            result = true;
        }

        return result;
    }
};

pub const Token = struct {
    tag: Tag,
    from: Index,
    to: Index,

    pub const Tag = enum(Index) {
        // operators
        @"+",
        @"+=",
        @"++",
        @" + ",
        @"<",
        @"<=",
        @" < ",
        @">",
        @">=",
        @" > ",
        @"!",
        @"!=",
        @"-",
        @"->",
        @"-=",
        @" - ",
        @".",
        @"..",
        @"..=",
        @"...",
        @":",
        @"::",
        @":-",
        @"*",
        @"*=",
        @" * ",
        @"/",
        @"/=",
        @" / ",
        @"%",
        @"%=",
        @" % ",
        @"=",
        @"=>",
        @"==",
        @"==>",
        @"~",
        @"~>",
        @"|",
        @"?",
        @"\\",
        @"#",
        @"&",
        @"[",
        @"]",
        @"(",
        @")",
        @"{",
        @"}",
        @",",
        @"'",
        @";",
        @"^",
        @"$",
        @"@",
        @"_",

        // permitive literals
        str,
        int,
        real,

        // keywords
        k_and,
        k_any,
        k_Any,
        k_as,
        k_assert,
        k_asume,
        k_async,
        k_await,
        k_bool,
        k_break,
        k_case,
        k_comptime,
        k_const,
        k_decreases,
        k_define,
        k_derive,
        k_do,
        k_dyn,
        k_effect,
        k_else,
        k_ensures,
        k_enum,
        k_error,
        k_exists,
        k_extern,
        k_false,
        k_fn,
        k_Fn,
        k_FnMut,
        k_FnOnce,
        k_for,
        k_forall,
        k_ghost,
        k_handle,
        k_if,
        k_impl,
        k_in,
        k_inline,
        k_invariant,
        k_is,
        k_lemma,
        k_let,
        k_lifting,
        k_match,
        k_move,
        k_mut,
        k_mutdyn,
        k_mutptr,
        k_mutref,
        k_newtype,
        k_noreturn,
        k_not,
        k_null,
        k_opaque,
        k_opens,
        k_or,
        k_pcfn,
        k_perform,
        k_predicate,
        k_ptr,
        k_pub,
        k_pure,
        k_ref,
        k_refines,
        k_requires,
        k_resume,
        k_return,
        k_returns,
        k_self,
        k_Self,
        k_static,
        k_struct,
        k_test,
        k_trait,
        k_true,
        k_tuple,
        k_typealias,
        k_undefined,
        k_union,
        k_unit,
        k_unreachable,
        k_unsafe,
        k_void,
        k_when,
        k_while,

        // others
        id,
        comment,
        invalid,
        sof,
        eof,
    };

    const keywords = std.StaticStringMap(Tag).initComptime(.{
        .{ "and", .k_and },
        .{ "any", .k_any },
        .{ "Any", .k_Any },
        .{ "as", .k_as },
        .{ "assert", .k_assert },
        .{ "asume", .k_asume },
        .{ "async", .k_async },
        .{ "await", .k_await },
        .{ "bool", .k_bool },
        .{ "break", .k_break },
        .{ "case", .k_case },
        .{ "comptime", .k_comptime },
        .{ "const", .k_const },
        .{ "decreases", .k_decreases },
        .{ "define", .k_define },
        .{ "derive", .k_derive },
        .{ "do", .k_do },
        .{ "dyn", .k_dyn },
        .{ "effect", .k_effect },
        .{ "else", .k_else },
        .{ "ensures", .k_ensures },
        .{ "enum", .k_enum },
        .{ "error", .k_error },
        .{ "exists", .k_exists },
        .{ "extern", .k_extern },
        .{ "false", .k_false },
        .{ "fn", .k_fn },
        .{ "Fn", .k_Fn },
        .{ "FnMut", .k_FnMut },
        .{ "FnOnce", .k_FnOnce },
        .{ "for", .k_for },
        .{ "forall", .k_forall },
        .{ "ghost", .k_ghost },
        .{ "handle", .k_handle },
        .{ "if", .k_if },
        .{ "impl", .k_impl },
        .{ "in", .k_in },
        .{ "inline", .k_inline },
        .{ "invariant", .k_invariant },
        .{ "is", .k_is },
        .{ "lemma", .k_lemma },
        .{ "let", .k_let },
        .{ "lifting", .k_lifting },
        .{ "match", .k_match },
        .{ "move", .k_move },
        .{ "mut", .k_mut },
        .{ "mutdyn", .k_mutdyn },
        .{ "mutptr", .k_mutptr },
        .{ "mutref", .k_mutref },
        .{ "newtype", .k_newtype },
        .{ "noreturn", .k_noreturn },
        .{ "not", .k_not },
        .{ "null", .k_null },
        .{ "opaque", .k_opaque },
        .{ "opens", .k_opens },
        .{ "or", .k_or },
        .{ "pcfn", .k_pcfn },
        .{ "perform", .k_perform },
        .{ "predicate", .k_predicate },
        .{ "ptr", .k_ptr },
        .{ "pub", .k_pub },
        .{ "pure", .k_pure },
        .{ "ref", .k_ref },
        .{ "refines", .k_refines },
        .{ "requires", .k_requires },
        .{ "resume", .k_resume },
        .{ "return", .k_return },
        .{ "returns", .k_returns },
        .{ "self", .k_self },
        .{ "Self", .k_Self },
        .{ "static", .k_static },
        .{ "struct", .k_struct },
        .{ "test", .k_test },
        .{ "trait", .k_trait },
        .{ "true", .k_true },
        .{ "tuple", .k_tuple },
        .{ "typealias", .k_typealias },
        .{ "undefined", .k_undefined },
        .{ "union", .k_union },
        .{ "unit", .k_unit },
        .{ "unreachable", .k_unreachable },
        .{ "unsafe", .k_unsafe },
        .{ "void", .k_void },
        .{ "when", .k_when },
        .{ "while", .k_while },
    });
};

inline fn token(tag: Token.Tag, from: Index, to: Index) Token {
    return .{ .tag = tag, .from = from, .to = to };
}

const Index = @import("../base/types.zig").Index;
const std = @import("std");

pub fn lex(gpa: std.mem.Allocator, src: []const u8) std.MultiArrayList(Token) {
    var lexer = Lexer.init(src);
    var result = std.MultiArrayList(Token){};
    result.append(gpa, token(.sof, 0, 0)) catch unreachable;
    while (true) {
        const t = lexer.next();
        result.append(gpa, t) catch unreachable;
        if (t.tag == .eof) break;
    }
    return result;
}

File: src/main/scala/lex/lex.scala
package lex

enum LexError {
  case UnexpectedEOF
  case InvalidChar
  case InvalidString
  case InvalidIdentifier
}

class Lexer(src: String) {
  var cursor: Int = 0
  // todo err: Unit = () // Scala Unit is like void, and defaults to ()

  private def token(tag: Tag, from: Int, to: Int): Token =
    Token(tag, from, to)

  def next(): Token = {
    if (cursor >= src.length) return token(Tag.eof, src.length, src.length)

    var oldCursor = cursor

    if (cursor >= src.length) return token(Tag.eof, src.length, src.length)

    val c = src(cursor)

    c match {
      case ' ' | '\t' | '\n' | '\r' =>
        cursor += 1
        next()

      case '^' =>
        cursor += 1
        token(Tag.`^`, oldCursor, cursor)

      case '.' =>
        cursor += 1
        token(Tag.`.`, oldCursor, cursor)

      case '\'' =>
        if (cursor + 2 < src.length && src(cursor + 2) == '\'') {
          // 'n' => char
          cursor += 3
          token(Tag.char, oldCursor, cursor)
        } else if (
          cursor + 3 < src.length && src(cursor + 1) == '\\' && src(
            cursor + 3
          ) == '\''
        ) {
          // '\n' => char
          cursor += 4
          token(Tag.char, oldCursor, cursor)
        } else {
          // '   { ANY } => macro_content
          // TODO: 实现宏内容处理
          cursor += 1
          token(Tag.`'`, oldCursor, cursor)
        }

      case '@' =>
        cursor += 1
        token(Tag.`@`, oldCursor, cursor)

      case '\\' =>
        cursor += 1
        token(Tag.`\\`, oldCursor, cursor)

      case '&' =>
        cursor += 1
        token(Tag.`&`, oldCursor, cursor)

      case '(' =>
        cursor += 1
        token(Tag.`(`, oldCursor, cursor)

      case ')' =>
        cursor += 1
        token(Tag.`)`, oldCursor, cursor)

      case '[' =>
        cursor += 1
        token(Tag.`[`, oldCursor, cursor)

      case ']' =>
        cursor += 1
        token(Tag.`]`, oldCursor, cursor)

      case ';' =>
        cursor += 1
        token(Tag.`;`, oldCursor, cursor)

      case ',' =>
        cursor += 1
        token(Tag.`,`, oldCursor, cursor)

      case '?' =>
        cursor += 1
        token(Tag.`?`, oldCursor, cursor)

      case '~' =>
        if (peek('>')) {
          cursor += 2
          token(Tag.`~>`, oldCursor, cursor)
        } else {
          cursor += 1
          token(Tag.`~`, oldCursor, cursor)
        }

      case '#' =>
        cursor += 1
        token(Tag.`#`, oldCursor, cursor)

      case '|' =>
        if (peek('>')) {
          cursor += 2
          token(Tag.`|>`, oldCursor, cursor)
        } else {
          cursor += 1
          token(Tag.`|`, oldCursor, cursor)
        }

      case '>' =>
        if (isSeparated()) {
          cursor += 2
          token(Tag.`_>_`, oldCursor, cursor)
        } else if (peek('=')) {
          cursor += 2
          token(Tag.`>=`, oldCursor, cursor)
        } else {
          cursor += 1
          token(Tag.`>`, oldCursor, cursor)
        }

      case '<' =>
        if (isSeparated()) {
          cursor += 2
          token(Tag.`_<_`, oldCursor, cursor)
        } else if (peek('=')) {
          cursor += 2
          token(Tag.`<=`, oldCursor, cursor)
        } else {
          cursor += 1
          token(Tag.`<`, oldCursor, cursor)
        }

      case '+' =>
        if (isSeparated()) {
          cursor += 2
          token(Tag.`_+_`, oldCursor, cursor)
        } else if (peek('=')) {
          cursor += 2
          token(Tag.`+=`, oldCursor, cursor)
        } else if (peek('+')) {
          cursor += 2
          token(Tag.`++`, oldCursor, cursor)
        } else {
          cursor += 1
          token(Tag.`+`, oldCursor, cursor)
        }

      case '-' =>
        if (isSeparated()) {
          cursor += 2
          token(Tag.`_-_`, oldCursor, cursor)
        } else if (peek('=')) {
          cursor += 2
          token(Tag.`-=`, oldCursor, cursor)
        } else if (peek('>')) {
          cursor += 2
          token(Tag.`->`, oldCursor, cursor)
        } else if (peek('-')) {
          // 处理单行注释
          cursor += 2
          while (cursor < src.length && src(cursor) != '\n') {
            cursor += 1
          }
          // 不跳过换行符，让下一个token处理它
          token(Tag.comment, oldCursor, cursor)
        } else {
          cursor += 1
          token(Tag.`-`, oldCursor, cursor)
        }

      case '!' =>
        if (peek('=')) {
          cursor += 2
          token(Tag.`!=`, oldCursor, cursor)
        } else {
          cursor += 1
          token(Tag.`!`, oldCursor, cursor)
        }

      case '=' =>
        if (peek('=')) {
          if (cursor + 2 < src.length && src(cursor + 2) == '>') {
            cursor += 3
            token(Tag.`==>`, oldCursor, cursor)
          } else {
            cursor += 2
            token(Tag.`==`, oldCursor, cursor)
          }
        } else if (peek('>')) {
          cursor += 2
          token(Tag.`=>`, oldCursor, cursor)
        } else {
          cursor += 1
          token(Tag.`=`, oldCursor, cursor)
        }

      case '%' =>
        if (isSeparated()) {
          cursor += 2
          token(Tag.`_%_`, oldCursor, cursor)
        } else if (peek('=')) {
          cursor += 2
          token(Tag.`%=`, oldCursor, cursor)
        } else {
          cursor += 1
          token(Tag.`%`, oldCursor, cursor)
        }

      case '{' =>
        if (peek('-')) {
          // 处理多行注释 {- ... -}
          cursor += 2
          var nesting = 1
          while (cursor < src.length && nesting > 0) {
            if (cursor + 1 < src.length) {
              if (src(cursor) == '{' && src(cursor + 1) == '-') {
                nesting += 1
                cursor += 2
              } else if (src(cursor) == '-' && src(cursor + 1) == '}') {
                nesting -= 1
                cursor += 2
              } else {
                cursor += 1
              }
            } else {
              // 如果到达文件末尾但注释未闭合，返回无效token
              return token(Tag.invalid, oldCursor, cursor)
            }
          }
          token(Tag.comment, oldCursor, cursor)
        } else {
          cursor += 1
          token(Tag.`{`, oldCursor, cursor)
        }

      case '}' =>
        cursor += 1
        token(Tag.`}`, oldCursor, cursor)

      case ':' =>
        if (peek('-')) {
          cursor += 2
          token(Tag.`:-`, oldCursor, cursor)
        } else if (peek(':')) {
          cursor += 2
          token(Tag.`::`, oldCursor, cursor)
        } else if (peek('~')) {
          cursor += 2
          token(Tag.`:~`, oldCursor, cursor)
        } else {
          cursor += 1
          token(Tag.`:`, oldCursor, cursor)
        }

      case '*' =>
        if (isSeparated()) {
          cursor += 2
          token(Tag.`_*_`, oldCursor, cursor)
        } else if (peek('=')) {
          cursor += 2
          token(Tag.`*=`, oldCursor, cursor)
        } else {
          cursor += 1
          token(Tag.`*`, oldCursor, cursor)
        }

      case '/' =>
        if (isSeparated()) {
          cursor += 2
          token(Tag.`_/_`, oldCursor, cursor)
        } else if (peek('=')) {
          cursor += 2
          token(Tag.`/=`, oldCursor, cursor)
        } else {
          cursor += 1
          token(Tag.`/`, oldCursor, cursor)
        }

      case '$' =>
        cursor += 1
        token(Tag.`$`, oldCursor, cursor)

      case c if isIdentifierStart(c) =>
        recognizeId()

      case '`' =>
        recognizeArbitaryId()

      case c if c.isDigit =>
        recognizeDigit()

      case '"' =>
        recognizeStr()

      case _ =>
        cursor += 1
        token(Tag.invalid, oldCursor, cursor)
    }
  }

  private def isIdentifierStart(c: Char): Boolean = {
    c.isLetter || c == '_'
  }

  private def isIdentifierPart(c: Char): Boolean = {
    c.isLetterOrDigit || c == '_'
  }

  private def peek(char: Char): Boolean = {
    cursor + 1 < src.length && src(cursor + 1) == char
  }

  private def isSeparated(): Boolean = {
    cursor > 0 && cursor + 1 < src.length &&
    src(cursor - 1) == ' ' && src(cursor + 1) == ' '
  }

  private def recognizeStr(): Token = {
    val oldCursor = cursor
    cursor += 1 // 跳过开始的双引号 "

    while (cursor < src.length) {
      val c = src(cursor)

      if (c == '"' && (cursor == 0 || src(cursor - 1) != '\\')) {
        cursor += 1 // 包含结束的双引号
        return token(Tag.str, oldCursor, cursor)
      }

      cursor += 1
    }

    // 如果没有找到结束的双引号，返回无效token
    token(Tag.invalid, oldCursor, cursor)
  }

  private def recognizeId(): Token = {
    val oldCursor = cursor
    
    // 读取标识符字符
    while (cursor < src.length && isIdentifierPart(src(cursor))) {
      cursor += 1
    }
    
    // 提取标识符并检查是否是关键字
    val identifier = src.substring(oldCursor, cursor)
    Tag.keywords.get(identifier) match {
      case Some(kw) => token(kw, oldCursor, cursor)
      case None     => token(Tag.id, oldCursor, cursor)
    }
  }

  // 23 => int(23)
  // 23.0 => real(23.0)
  private def recognizeDigit(): Token = {
    val oldCursor = cursor
    var isReal = false

    // 读取整数部分
    while (cursor < src.length && (src(cursor).isDigit || src(cursor) == '_')) {
      cursor += 1
    }

    // 检查是否有小数点和小数部分
    if (cursor < src.length && src(cursor) == '.') {
      if (cursor + 1 < src.length && src(cursor + 1).isDigit) {
        isReal = true
        cursor += 1 // 跳过小数点
        
        // 读取小数部分
        while (cursor < src.length && (src(cursor).isDigit || src(cursor) == '_')) {
          cursor += 1
        }
      }
    }

    if (isReal) {
      token(Tag.real, oldCursor, cursor)
    } else {
      token(Tag.int, oldCursor, cursor)
    }
  }

  private def recognizeArbitaryId(): Token = {
    val oldCursor = cursor
    cursor += 1 // 跳过开始的反引号 `

    while (cursor < src.length) {
      val c = src(cursor)

      if (c == '`') {
        cursor += 1 // 包含结束的反引号
        return token(Tag.id, oldCursor, cursor)
      }

      cursor += 1
    }

    // 如果没有找到结束的反引号，返回无效token
    token(Tag.invalid, oldCursor, cursor)
  }

  private def seperated(at: Int): Boolean = {
    if (src.length <= at + 1 || at < 1) false
    else if (src.charAt(at + 1) == ' ' && src.charAt(at - 1) == ' ') true
    else false
  }

  private def eatChar(char: Char): Boolean = {
    var result: Boolean = false

    if (src.length <= cursor + 1) return false

    if (src.charAt(cursor + 1) == char) {
      cursor += 1
      result = true
    }

    result
  }
}
